packages/flutter/lib/src/material/mergeable_material.dart

```diff
@@ -0,0 +1,562 @@
+  /// [MergeableMaterial].
+  MaterialSlice({
+    LocalKey key,
+    this.child
+  }) : super(key);
+
+  /// The widget below this widget in the tree.
+  final Widget child;
+
+  @override
+  String toString() {
+    return 'MergeableSlice(key: $key, child: $child)';
+  }
+}
+
+/// A class that represents a gap within [MergeableMaterial].
+///
+/// All [MaterialGap] objects need a [LocalKey].
+class MaterialGap extends MergeableMaterialItem {
+  /// Creates a Material gap with a given size.
+  MaterialGap({
+    LocalKey key,
+    this.size: 16.0
+  }) : super(key);
+
+  /// The size of the gap, irrespective of the axis used.
+  final double size;
+
+  @override
+  String toString() {
+    return 'MaterialGap(key: $key, child: $size)';
+  }
+}
+
+/// A widget useful for animating the merging and separating of [Material]
+/// slices. The widget takes [MergeableMaterialItem] children, uses a block
+/// layout algorithm, and separates [MaterialSlice] wherever it sees a
+/// [MaterialGap]. Animations work by figuring out where new gaps were
+/// added/removed when the widget gets rebuilt by checking the children's keys,
+/// which is why keys on children are mandatory.
+///
+/// The addition of extra slices is not animated and neither is the adding of
+/// any extra items at the end of the children list.
+class MergeableMaterial extends StatefulWidget {
+  /// Creates a mergeable Material list of items.
+  MergeableMaterial({
+    Key key,
+    this.mainAxis: Axis.vertical,
+    this.elevation: 2,
+    this.children: const <MergeableMaterialItem>[]
+  }) : super(key: key);
+
+  /// The children of the [MergeableMaterial].
+  final List<MergeableMaterialItem> children;
+
+  /// The main layout axis.
+  final Axis mainAxis;
+
+  /// The elevation of all the [Material] slices.
+  final int elevation;
+
+  @override
+  String toString() {
+    return 'MergeableMaterial('
+      'key: $key, mainAxis: $mainAxis, '
+      'elevation: $elevation, children: $children'
+    ')';
+  }
+
+  @override
+  _MergeableMaterialState createState() {
+    return new _MergeableMaterialState();
+  }
+}
+
+class _AnimationTuple {
+  _AnimationTuple({
+    this.controller,
+    this.startAnimation,
+    this.endAnimation,
+    this.gapAnimation,
+    this.gapStart: 0.0
+  });
+
+  final AnimationController controller;
+  final CurvedAnimation startAnimation;
+  final CurvedAnimation endAnimation;
+  final CurvedAnimation gapAnimation;
+  double gapStart;
+}
+
+class _MergeableMaterialState extends State<MergeableMaterial> {
+  List<MergeableMaterialItem> _children;
+  Map<LocalKey, _AnimationTuple> _animationTuples;
+
+  void _initGap(MaterialGap gap) {
+    final AnimationController controller = new AnimationController(
+      duration: kThemeAnimationDuration
+    );
+
+    final CurvedAnimation startAnimation = new CurvedAnimation(
+      parent: controller,
+      curve: Curves.ease,
+      reverseCurve: Curves.ease
```

Comment:
(so in this case, no need to give reverseCurve, since that defaults to the given curve)